In [5]:
import MyPyEigen as pe
import numpy as np
import timeit

def benchmark_function(func, args, number=100, preheat=10):
    """
    Benchmark 'func' with arguments 'args'. Preheat the cache by calling the function
    'preheat' times. Then, call 'func' 'number' times and return the average execution time in seconds.
    """
    for _ in range(preheat):
        func(*args)
    return timeit.timeit(lambda: func(*args), number=number) / number

# Ensure reproducibility
np.random.seed(12345)

# Define test sizes (lengths for the 1D FFT input)
sizes = [10, 50, 100, 500, 1000, 2000, 5000, 10000, 20000, 50000, 100000, 200000, 500000]
results_fft = {"numpy": [], "MyPyEigen": []}

for size in sizes:
    # Create a 1D array of the given size (dtype: float64)
    x = np.random.rand(size).astype(np.float64)
    
    # Benchmark NumPy's FFT (using np.fft.fft)
    t_numpy = benchmark_function(np.fft.fft, (x,), number=100)
    results_fft["numpy"].append((size, t_numpy))
    
    # Benchmark MyPyEigen's FFT (using pe.fft)
    t_pe = benchmark_function(pe.fft, (x,), number=100)
    results_fft["MyPyEigen"].append((size, t_pe))

print("FFT Benchmark:")
print("Size\tNumPy (sec)\tMyPyEigen (sec)")
for size, t in results_fft["numpy"]:
    t_pe = [item[1] for item in results_fft["MyPyEigen"] if item[0] == size][0]
    print(f"{size}\t{t:.9f}\t{t_pe:.9f}")


FFT Benchmark:
Size	NumPy (sec)	MyPyEigen (sec)
10	0.000004652	0.000001740
50	0.000004967	0.000002290
100	0.000005365	0.000002746
500	0.000008058	0.000006335
1000	0.000013530	0.000011260
2000	0.000020867	0.000020802
5000	0.000121157	0.000061976
10000	0.000189841	0.000113537
20000	0.000224010	0.000204683
50000	0.000598528	0.000564668
100000	0.001451865	0.001233335
200000	0.003095395	0.002721594
500000	0.007363444	0.007539679


### FFT Benchmark
| Size   | NumPy (sec)   | MyPyEigen (sec) | Speedup (MyPyEigen vs NumPy) |
|--------|---------------|------------------|------------------------------|
| 10     | 0.000004652   | 0.000001740      | **2.67x** (Faster)           |
| 50     | 0.000004967   | 0.000002290      | **2.17x** (Faster)           |
| 100    | 0.000005365   | 0.000002746      | **1.95x** (Faster)           |
| 500    | 0.000008058   | 0.000006335      | **1.27x** (Faster)           |
| 1000   | 0.000013530   | 0.000011260      | **1.20x** (Faster)           |
| 2000   | 0.000020867   | 0.000020802      | **1.00x** (Similar)          |
| 5000   | 0.000121157   | 0.000061976      | **1.95x** (Faster)           |
| 10000  | 0.000189841   | 0.000113537      | **1.67x** (Faster)           |
| 20000  | 0.000224010   | 0.000204683      | **1.09x** (Faster)           |
| 50000  | 0.000598528   | 0.000564668      | **1.06x** (Faster)           |
| 100000 | 0.001451865   | 0.001233335      | **1.18x** (Faster)           |
| 200000 | 0.003095395   | 0.002721594      | **1.14x** (Faster)           |
| 500000 | 0.007363444   | 0.007539679      | **0.98x** (Similar)          |